webapp/tests/karma/ts/services/rules-engine.service.spec.ts

```diff
@@ -0,0 +1,619 @@
+    expect(actual.length).to.eq(1);
+    expect(actual[0]).to.nested.include({
+      _id: 'taskdoc',
+      'emission.other': true,
+    });
+    expect(telemetryService.record.callCount).to.equal(4);
+    expect(telemetryService.record.args[1]).to.deep.equal(['rules-engine:tasks:dirty-contacts', 3]);
+    expect(telemetryService.record.args[2][0]).to.equal('rules-engine:ensureTaskFreshness:cancel');
+    expect(telemetryService.record.args[3][0]).to.equal('rules-engine:tasks:all-contacts');
+  });
+
+  it('fetchTaskDocsFor() should fetch task docs', async () => {
+    const taskDoc = JSON.parse(JSON.stringify(sampleTaskDoc));
+    const contactIds = ['a', 'b', 'c'];
+    fetchTasksResult = sinon.stub().resolves([taskDoc]);
+    rulesEngineCoreStubs.getDirtyContacts.returns(['a', 'b']);
+    service = TestBed.inject(RulesEngineService);
+
+    const actual = await service.fetchTaskDocsFor(contactIds);
+
+    expect(rulesEngineCoreStubs.fetchTasksFor.callCount).to.eq(1);
+    expect(rulesEngineCoreStubs.fetchTasksFor.args[0][0]).to.eq(contactIds);
+    expect(actual.length).to.eq(1);
+    expect(actual[0]).to.nested.include({
+      _id: 'taskdoc',
+      'emission.other': true,
+    });
+    expect(telemetryService.record.callCount).to.equal(3);
+    expect(telemetryService.record.args[1]).to.deep.equal(['rules-engine:tasks:dirty-contacts', 2]);
+    expect(telemetryService.record.args[2][0]).to.equal('rules-engine:tasks:some-contacts');
+  });
+
+  it('fetchTargets() should send correct range to Rules Engine Core when getting targets', async () => {
+    fetchTargetsResult = sinon.stub().resolves([]);
+    service = TestBed.inject(RulesEngineService);
+
+    const actual = await service.fetchTargets();
+
+    expect(actual).to.deep.eq([]);
+    expect(rulesEngineCoreStubs.fetchTargets.callCount).to.eq(1);
+    expect(rulesEngineCoreStubs.fetchTargets.args[0][0]).to.have.keys('start', 'end');
+  });
+
+  it('should ensure freshness of tasks and targets', async () => {
+    rulesEngineCoreStubs.getDirtyContacts.returns(Array.from({ length: 20 }).map(i => i));
+    const clock = sinon.useFakeTimers(1000);
+    service = TestBed.inject(RulesEngineService);
+
+    await service.isEnabled();
+    clock.tick(500 * 1000);
+    await service.isEnabled(); // to resolve promises
+    
+    expect(rulesEngineCoreStubs.fetchTasksFor.callCount).to.eq(1);
+    expect(rulesEngineCoreStubs.fetchTargets.callCount).to.eq(1);
+    expect(telemetryService.record.callCount).to.equal(3);
+    expect(telemetryService.record.args[0][0]).to.equal('rules-engine:initialize');
+    expect(telemetryService.record.args[1]).to.deep.equal(['rules-engine:tasks:dirty-contacts', 20]);
+    expect(telemetryService.record.args[2]).to.deep.equal(['rules-engine:targets:dirty-contacts', 20]);
+  });
+
+  it('should ensure freshness of tasks only', async () => {
+    fetchTargetsResult = sinon.stub().resolves([]);
+    const clock = sinon.useFakeTimers(1000);
```

Comment:
I think you have to explictly call `clock.restore` to restore normal time. If you do it after assertions in this test, and the test fails, you might get other following tests to fail because time is not stored. 
If I need to mock time, I make sure to make `clock`  a top scope variable (but I call `useFakeTimers` in the test where I need it) and make sure to `clock && clock.restore()` in the `afterEach`.

Reply:
I see, thank you! Fixed